sklearn/cluster/_hierarchical.pyx

```diff
@@ -332,3 +332,98 @@ cdef class WeightedEdge:
                                               self.weight,
                                               self.a, self.b)
 
+
+################################################################################
+# Efficient labelling/conversion of MSTs to single linkage hierarchies
+
+cdef class UnionFind(object):
+
+    cdef np.ndarray parent_arr
+    cdef np.ndarray size_arr
+    cdef ITYPE_t next_label
+    cdef ITYPE_t *parent
+    cdef ITYPE_t *size
+
+    def __init__(self, N):
+        self.parent_arr = -1 * np.ones(2 * N - 1, dtype=ITYPE, order='C')
+        self.next_label = N
+        self.size_arr = np.hstack((np.ones(N, dtype=ITYPE),
+                                   np.zeros(N-1, dtype=ITYPE)))
+        self.parent = (<ITYPE_t *> self.parent_arr.data)
+        self.size = (<ITYPE_t *> self.size_arr.data)
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef void union(self, ITYPE_t m, ITYPE_t n):
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.parent[m] = self.next_label
+        self.parent[n] = self.next_label
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.next_label += 1
+
+        return
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef ITYPE_t fast_find(self, ITYPE_t n):
+        cdef ITYPE_t p
+        p = n
+        while self.parent_arr[n] != -1:
+            n = self.parent_arr[n]
+        # label up to the root
+        while self.parent_arr[p] != n:
+            p, self.parent_arr[p] = self.parent_arr[p], n
+        return n
+
+
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef np.ndarray[DTYPE_t, ndim=2] single_linkage_label(
+    np.ndarray[DTYPE_t, ndim=2] L):
+    """
+    Convert an linkage array or MST to a tree by labelling clusters at merges.
+    This is done by using a Union find structure to keep track of merges
+    efficiently.
+
+    Parameters
+    ----------
+
+    L: array of shape (n_samples - 1, 3)
+        The linkage array or MST where each row specifies two samples
+        to be merged and a distance or weight at which the merge occurs. This
+         array is assumed to be sorted by the distance/weight.
+
+    Returns
+    -------
+
+    A tree in the format used by scipy.cluster.hierarchy.
+    """
+
+    cdef np.ndarray[DTYPE_t, ndim=2] result_arr
+    cdef DTYPE_t[:, ::1] result
+
+    cdef ITYPE_t N, a, aa, b, bb, index
+    cdef DTYPE_t delta
+
+    result_arr = np.zeros((L.shape[0], L.shape[1] + 1), dtype=DTYPE)
+    result = (<DTYPE_t[:L.shape[0], :4:1]> (
+        <DTYPE_t *> result_arr.data))
+    N = L.shape[0] + 1
+    U = UnionFind(N)
+
+    for index in range(L.shape[0]):
+
+        a = <ITYPE_t> L[index, 0]
+        b = <ITYPE_t> L[index, 1]
+        delta = L[index, 2]
+
+        aa, bb = U.fast_find(a), U.fast_find(b)
```

Comment:
I don't see why this should use unpacking syntax, while without it, we could benefit from a more self-documenting variable name like a_cluster, b_cluster